sdk/client/src/broadcast_client.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{error::WaitForTransactionError, views::{AccountView, MetadataView, TransactionView}, Client, Error, Response, Result, MethodRequest, MethodResponse};
+use diem_crypto::hash::CryptoHash;
+use diem_types::{
+    account_address::AccountAddress,
+    transaction::{SignedTransaction, Transaction},
+};
+
+use crate::views::{CurrencyInfoView, EventView, StateProofView, AccountStateWithProofView};
+use futures::future::join_all;
+use rand::seq::{SliceChooseIter, SliceRandom};
+use std::time::Duration;
+
+#[derive(Clone, Debug)]
+struct BroadcastingClient {
+    pub clients: Vec<Client>,
+    pub num_parallel_requests: usize,
+}
+
+impl BroadcastingClient {
+    pub fn new(clients: Vec<Client>, num_parallel_requests: usize) -> Result<Self, String> {
+        if num_parallel_requests < 1 {
+            return Err("num_parallel_requests should be >= 1".into());
+        }
+        if num_parallel_requests > clients.len() {
+            return Err(format!(
+                "num_parallel_requests({}) should be <= length of clients({})",
+                num_parallel_requests,
+                clients.len()
+            ));
+        }
+        Ok(Self {
+            clients,
+            num_parallel_requests,
+        })
+    }
+
+    pub fn random_clients(&self) -> SliceChooseIter<'_, [Client], Client> {
+        self.clients
+            .choose_multiple(&mut rand::thread_rng(), self.num_parallel_requests)
+    }
+
+    pub async fn batch(&self, requests: Vec<MethodRequest>) -> Result<Vec<Vec<Result<Response<MethodResponse>>>>> {
+        let futures = self.random_clients().map(|client| client.batch(requests.clone()));
+        let mut results = join_all(futures).await;
+        if results.is_empty() {
+            return Err(Error::unknown("Expected at least one response, found none"));
+        }
+        // All of them returned an error, so return the first error
+        if results.iter().all(|x| x.is_err()) {
+            if let Err(e) = results.swap_remove(0) {
+                return Err(e);
+            }
+        }
+        Ok(results.into_iter().filter_map(|x| x.ok()).collect())
+    }
+
+    pub async fn submit(&self, txn: &SignedTransaction) -> Result<Vec<Response<()>>> {
```

Comment:
At the end of the day we should only ever return 1 response, not all of the responses. So the logic in collect results should probably try to select just 1 to return.